testing/test_runner.py

```diff
@@ -762,6 +765,27 @@ def test_importorskip_imports_last_module_part() -> None:
     assert os.path == ospath
 
 
+def test_importorskip_importError_Exception() -> None:
+    ## Mocking the import function to raise a importError
+    realimport = builtins.__import__
+
+    def myimport(
+        name: str,
+        globals: Mapping[str, object] | None = None,
+        locals: Mapping[str, object] | None = None,
+        fromlist: Sequence[str] = (),
+        level: int = 0,
+    ) -> types.ModuleType:
+        raise ImportError
+
+    builtins.__import__ = myimport
+
+    with pytest.raises(ImportError):
+        pytest.importorskip("abcdefghi")
+
+    builtins.__import__ = realimport
```

Comment:
Instead of mocking `builtins.__import__`, which seems dangerous, let's instead use `pytester` and generate actual modules and try to import them;

```suggestion
def test_importorskip_module_not_found_error(pytester: Pytester) -> None:
    """
    importorskip() will only skip modules if they were not actually found via ``ModuleNotFoundError``, instead of catching the more generic ``ImportError``.
    
    #11523
    """
    fn = pytester.makepyfile("raise ImportError('some specific problem')")
    pytester.syspathinsert()
    
    with pytest.raises(ImportError, match="some specific problem"):
        pytest.importorskip(fn.stem)
        
    # User might decide to skip on any ImportError.    
    pytest.importorskip(fn.stem, exc=ImportError)  # passes
```